pkg/daemon/writer.go

```diff
@@ -126,24 +127,42 @@ func (nw *clusterNodeWriter) Run(stop <-chan struct{}) {
 		case <-stop:
 			return
 		case msg := <-nw.writer:
-			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos)
+			r := implSetNodeAnnotations(nw.client, nw.lister, nw.nodeName, msg.annos, msg.annosToDelete)
 			msg.responseChannel <- r
 		}
 	}
 }
 
 // SetDone sets the state to Done.
-func (nw *clusterNodeWriter) SetDone(dcAnnotation string) error {
+func (nw *clusterNodeWriter) SetDone(state *stateAndConfigs) error {
 	annos := map[string]string{
 		constants.MachineConfigDaemonStateAnnotationKey: constants.MachineConfigDaemonStateDone,
-		constants.CurrentMachineConfigAnnotationKey:     dcAnnotation,
+		constants.CurrentMachineConfigAnnotationKey:     state.currentConfig.GetName(),
 		// clear out any Degraded/Unreconcilable reason
 		constants.MachineConfigDaemonReasonAnnotationKey: "",
 	}
+
+	// If current image is not empty, update the annotation.
+	if state.currentImage != "" {
+		annos[constants.CurrentImageAnnotationKey] = state.currentImage
+	}
+
+	annosToDelete := []string{}
+	// If current image is empty, delete the annotation, if it exists.
+	if state.currentImage == "" {
+		annosToDelete = append(annosToDelete, constants.CurrentImageAnnotationKey)
+	}
+
+	// If desired image is empty, delete the annotation, if it exists.
+	if state.desiredImage == "" {
```

Comment:
This logic seems fine to me, although I think there isn't really ever a case where we can delete the desiredImage annotation (if it exists already) right? Since we can't rollback?

Most of the time this would be a no-op which is fine

Reply:
This sets us up for the rollback scenario although perhaps it might be unnecessary right now. I don't think it would cause any side-effects for normal MachineConfig updates since those annotations shouldn't be present anyway.